File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        
        if not root: 
            return 0
        
        def pathDiff(node, max_value, min_value):
            if not node:
                return max_value - min_value
            
            max_value = max(max_value, node.val)
            min_value = min(min_value, node.val)
            
            left = pathDiff(node.left, max_value, min_value)
            right = pathDiff(node.right, max_value, min_value)
            
            return max(left, right)
        
        return pathDiff(root, root.val, root.val)

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pathDiff(TreeNode* root, int max_value, int min_value) {
        if(root == NULL)
            return max_value - min_value;
            
        max_value = max(max_value, root->val);
        min_value = min(min_value, root->val);
        
        int left = pathDiff(root->left, max_value, min_value);
        int right = pathDiff(root->right, max_value, min_value);
        
        return max(left, right);
    }
    
    int maxAncestorDiff(TreeNode* root) {
        return pathDiff(root, root->val, root->val);
    }
};

